File: packages/shield-sdk/src/shield.ts
import type { Scanner, ScanResult, ShieldConfig, ScanOptions, RiskFactor, Entity, Severity } from './types.js';
import { PromptInjectionScanner } from './scanners/prompt-injection.js';
import { SecretsScanner } from './scanners/secrets.js';
import { PIIScanner } from './scanners/pii.js';
import { ToxicityScanner } from './scanners/toxicity.js';

/**
 * Shield - Main security facade for LLM applications
 *
 * @example
 * ```typescript
 * // Using presets
 * const shield = Shield.standard();
 * const result = await shield.scanPrompt("User input");
 *
 * // Custom configuration
 * const shield = Shield.builder()
 *   .addInputScanner(new SecretsScanner())
 *   .addInputScanner(new PIIScanner())
 *   .withParallelExecution(true)
 *   .build();
 * ```
 */
export class Shield {
  private inputScanners: Scanner[];
  private outputScanners: Scanner[];
  private config: ShieldConfig;

  private constructor(config: ShieldConfig, inputScanners: Scanner[], outputScanners: Scanner[]) {
    this.config = config;
    this.inputScanners = inputScanners;
    this.outputScanners = outputScanners;
  }

  /**
   * Create a Shield with strict security settings
   * Maximum security for regulated industries (banking, healthcare)
   */
  static strict(): Shield {
    return new Shield(
      {
        preset: 'strict',
        shortCircuitThreshold: 0.7,
        parallelExecution: false,
      },
      [
        new PromptInjectionScanner(),
        new SecretsScanner(),
        new PIIScanner(),
        new ToxicityScanner(),
      ],
      [
        new SecretsScanner(),
        new PIIScanner(),
      ]
    );
  }

  /**
   * Create a Shield with standard security settings
   * Balanced security for general-purpose applications (recommended)
   */
  static standard(): Shield {
    return new Shield(
      {
        preset: 'standard',
        shortCircuitThreshold: 0.9,
        parallelExecution: true,
        maxConcurrent: 4,
      },
      [
        new PromptInjectionScanner(),
        new SecretsScanner(),
        new PIIScanner({ piiTypes: ['email', 'ssn', 'credit-card'] }),
      ],
      [
        new SecretsScanner(),
        new PIIScanner({ piiTypes: ['email', 'ssn', 'credit-card'] }),
      ]
    );
  }

  /**
   * Create a Shield with permissive security settings
   * Minimal security for development/testing
   */
  static permissive(): Shield {
    return new Shield(
      {
        preset: 'permissive',
        shortCircuitThreshold: 1.0,
        parallelExecution: true,
      },
      [
        new SecretsScanner({ secretTypes: ['aws', 'private-key'] }),
      ],
      []
    );
  }

  /**
   * Create a ShieldBuilder for custom configuration
   */
  static builder(): ShieldBuilder {
    return new ShieldBuilder();
  }

  /**
   * Scan a prompt before sending to LLM
   */
  async scanPrompt(text: string, options?: ScanOptions): Promise<ScanResult> {
    return this.runScanners(text, this.inputScanners, options);
  }

  /**
   * Scan LLM output before returning to user
   */
  async scanOutput(text: string, options?: ScanOptions): Promise<ScanResult> {
    return this.runScanners(text, this.outputScanners, options);
  }

  /**
   * Scan both prompt and output in sequence
   */
  async scanPromptAndOutput(
    prompt: string,
    output: string,
    options?: ScanOptions
  ): Promise<{ promptResult: ScanResult; outputResult: ScanResult }> {
    const promptResult = await this.scanPrompt(prompt, options);

    // Short-circuit if prompt is invalid
    if (!promptResult.isValid && promptResult.riskScore >= (this.config.shortCircuitThreshold ?? 0.9)) {
      return {
        promptResult,
        outputResult: {
          isValid: false,
          riskScore: 0,
          sanitizedText: '',
          entities: [],
          riskFactors: [{
            category: 'prompt-injection',
            description: 'Output scan skipped due to invalid prompt',
            severity: 'none',
            confidence: 1.0,
          }],
          severity: 'none',
          metadata: { skipped: 'true' },
          durationMs: 0,
        },
      };
    }

    const outputResult = await this.scanOutput(output, options);
    return { promptResult, outputResult };
  }

  /**
   * Scan multiple texts in batch
   */
  async scanBatch(texts: string[], options?: ScanOptions): Promise<ScanResult[]> {
    if (this.config.parallelExecution) {
      return Promise.all(texts.map(text => this.scanPrompt(text, options)));
    }

    const results: ScanResult[] = [];
    for (const text of texts) {
      results.push(await this.scanPrompt(text, options));
    }
    return results;
  }

  private async runScanners(
    text: string,
    scanners: Scanner[],
    options?: ScanOptions
  ): Promise<ScanResult> {
    if (scanners.length === 0) {
      return {
        isValid: true,
        riskScore: 0,
        sanitizedText: text,
        entities: [],
        riskFactors: [],
        severity: 'none',
        metadata: {},
        durationMs: 0,
      };
    }

    const startTime = performance.now();
    const processedText = options?.maxLength ? text.slice(0, options.maxLength) : text;

    let results: ScanResult[];

    if (this.config.parallelExecution) {
      const maxConcurrent = this.config.maxConcurrent ?? 4;
      results = await this.runParallel(processedText, scanners, maxConcurrent);
    } else {
      results = await this.runSequential(processedText, scanners);
    }

    // Merge results
    const mergedResult = this.mergeResults(processedText, results);
    mergedResult.durationMs = performance.now() - startTime;

    return mergedResult;
  }

  private async runSequential(text: string, scanners: Scanner[]): Promise<ScanResult[]> {
    const results: ScanResult[] = [];

    for (const scanner of scanners) {
      const result = await scanner.scan(text);
      results.push(result);

      // Short-circuit if threshold exceeded
      if (result.riskScore >= (this.config.shortCircuitThreshold ?? 0.9)) {
        break;
      }
    }

    return results;
  }

  private async runParallel(
    text: string,
    scanners: Scanner[],
    maxConcurrent: number
  ): Promise<ScanResult[]> {
    const results: ScanResult[] = [];

    for (let i = 0; i < scanners.length; i += maxConcurrent) {
      const batch = scanners.slice(i, i + maxConcurrent);
      const batchResults = await Promise.all(batch.map(s => s.scan(text)));
      results.push(...batchResults);

      // Check for short-circuit
      const maxRisk = Math.max(...batchResults.map(r => r.riskScore));
      if (maxRisk >= (this.config.shortCircuitThreshold ?? 0.9)) {
        break;
      }
    }

    return results;
  }

  private mergeResults(originalText: string, results: ScanResult[]): ScanResult {
    if (results.length === 0) {
      return {
        isValid: true,
        riskScore: 0,
        sanitizedText: originalText,
        entities: [],
        riskFactors: [],
        severity: 'none',
        metadata: {},
        durationMs: 0,
      };
    }

    const allEntities: Entity[] = [];
    const allRiskFactors: RiskFactor[] = [];
    let maxRiskScore = 0;
    let maxSeverity: Severity = 'none';

    const severityOrder: Severity[] = ['none', 'low', 'medium', 'high', 'critical'];

    for (const result of results) {
      allEntities.push(...result.entities);
      allRiskFactors.push(...result.riskFactors);

      if (result.riskScore > maxRiskScore) {
        maxRiskScore = result.riskScore;
      }

      const severityIndex = severityOrder.indexOf(result.severity);
      if (severityIndex > severityOrder.indexOf(maxSeverity)) {
        maxSeverity = result.severity;
      }
    }

    // Deduplicate entities by position
    const uniqueEntities = this.deduplicateEntities(allEntities);

    // Get sanitized text from the last result that modified it
    let sanitizedText = originalText;
    for (const result of results) {
      if (result.sanitizedText !== originalText) {
        sanitizedText = result.sanitizedText;
      }
    }

    return {
      isValid: allRiskFactors.length === 0,
      riskScore: maxRiskScore,
      sanitizedText,
      entities: uniqueEntities,
      riskFactors: allRiskFactors,
      severity: maxSeverity,
      metadata: {},
      durationMs: 0,
    };
  }

  private deduplicateEntities(entities: Entity[]): Entity[] {
    const seen = new Set<string>();
    return entities.filter(entity => {
      const key = `${entity.start}-${entity.end}-${entity.entityType}`;
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }
}

/**
 * Builder for creating custom Shield configurations
 */
export class ShieldBuilder {
  private inputScanners: Scanner[] = [];
  private outputScanners: Scanner[] = [];
  private config: ShieldConfig = {
    parallelExecution: true,
    maxConcurrent: 4,
    shortCircuitThreshold: 0.9,
  };

  /**
   * Add an input scanner
   */
  addInputScanner(scanner: Scanner): this {
    this.inputScanners.push(scanner);
    return this;
  }

  /**
   * Add an output scanner
   */
  addOutputScanner(scanner: Scanner): this {
    this.outputScanners.push(scanner);
    return this;
  }

  /**
   * Set the short-circuit threshold
   */
  withShortCircuit(threshold: number): this {
    this.config.shortCircuitThreshold = threshold;
    return this;
  }

  /**
   * Enable or disable parallel execution
   */
  withParallelExecution(enabled: boolean): this {
    this.config.parallelExecution = enabled;
    return this;
  }

  /**
   * Set maximum concurrent scanners
   */
  withMaxConcurrent(max: number): this {
    this.config.maxConcurrent = max;
    return this;
  }

  /**
   * Build the Shield instance
   */
  build(): Shield {
    // Use private constructor via closure
    return (Shield as any)['strict']().constructor['call'](
      Object.create(Shield.prototype),
      this.config,
      this.inputScanners,
      this.outputScanners
    ) || new (Shield as any)(this.config, this.inputScanners, this.outputScanners);
  }
}

// Fix the build method to work properly
ShieldBuilder.prototype.build = function(this: ShieldBuilder): Shield {
  const shield = Shield.strict();
  (shield as any).config = (this as any).config;
  (shield as any).inputScanners = (this as any).inputScanners;
  (shield as any).outputScanners = (this as any).outputScanners;
  return shield;
};

File: crates/llm-shield-sdk/src/shield.rs
//! # Shield - Main API
//!
//! The Shield struct is the main entry point for the LLM Shield SDK.
//!
//! ## Features
//!
//! - Simple, intuitive API
//! - Async-first design
//! - Support for both prompt and output scanning
//! - Batch processing for high throughput
//! - Configurable parallel execution

use crate::builder::ShieldBuilder;
use crate::config::ShieldConfig;
use crate::error::{SdkError, SdkResult};
use crate::preset::Preset;
use futures::future::join_all;
use llm_shield_core::{ScanResult, Scanner, ScannerPipeline, Vault};
use std::sync::Arc;
use std::time::Instant;

/// Main entry point for the LLM Shield SDK
///
/// Shield provides a high-level API for scanning LLM prompts and outputs
/// for security threats, sensitive data, and policy violations.
///
/// ## Quick Start
///
/// ```rust,ignore
/// use llm_shield_sdk::prelude::*;
///
/// #[tokio::main]
/// async fn main() -> SdkResult<()> {
///     // Create a shield with standard security
///     let shield = Shield::standard()?;
///
///     // Scan a prompt
///     let result = shield.scan_prompt("Hello, how are you?").await?;
///
///     if result.is_valid {
///         println!("Prompt is safe!");
///     } else {
///         println!("Risk detected: {:.2}", result.risk_score);
///     }
///
///     Ok(())
/// }
/// ```
///
/// ## Custom Configuration
///
/// ```rust,ignore
/// let shield = Shield::builder()
///     .with_preset(Preset::Standard)
///     .with_short_circuit(0.9)
///     .with_parallel_execution(true)
///     .build()?;
/// ```
pub struct Shield {
    config: ShieldConfig,
    input_scanners: Vec<Arc<dyn Scanner>>,
    output_scanners: Vec<Arc<dyn Scanner>>,
    vault: Vault,
}

impl Shield {
    // ========================================================================
    // Constructors
    // ========================================================================

    /// Create a new Shield with the given configuration and scanners
    pub(crate) fn new(
        config: ShieldConfig,
        input_scanners: Vec<Arc<dyn Scanner>>,
        output_scanners: Vec<Arc<dyn Scanner>>,
    ) -> Self {
        Self {
            config,
            input_scanners,
            output_scanners,
            vault: Vault::new(),
        }
    }

    /// Create a builder for custom configuration
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let shield = Shield::builder()
    ///     .with_preset(Preset::Standard)
    ///     .build()?;
    /// ```
    pub fn builder() -> ShieldBuilder {
        ShieldBuilder::new()
    }

    /// Create a Shield with strict security (maximum protection)
    ///
    /// Recommended for regulated industries (banking, healthcare).
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let shield = Shield::strict()?;
    /// ```
    pub fn strict() -> SdkResult<Self> {
        ShieldBuilder::new().with_preset(Preset::Strict).build()
    }

    /// Create a Shield with standard security (balanced protection)
    ///
    /// Recommended for general-purpose applications.
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let shield = Shield::standard()?;
    /// ```
    pub fn standard() -> SdkResult<Self> {
        ShieldBuilder::new().with_preset(Preset::Standard).build()
    }

    /// Create a Shield with permissive security (minimal protection)
    ///
    /// Recommended for development and testing only.
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let shield = Shield::permissive()?;
    /// ```
    pub fn permissive() -> SdkResult<Self> {
        ShieldBuilder::new().with_preset(Preset::Permissive).build()
    }

    // ========================================================================
    // Scanning Methods
    // ========================================================================

    /// Scan a prompt before sending to an LLM
    ///
    /// ## Arguments
    ///
    /// * `prompt` - The user prompt to scan
    ///
    /// ## Returns
    ///
    /// A `ScanResult` containing:
    /// - `is_valid`: Whether the prompt passed all security checks
    /// - `risk_score`: Overall risk score (0.0 to 1.0)
    /// - `entities`: Detected entities (PII, secrets, etc.)
    /// - `risk_factors`: Contributing risk factors
    /// - `sanitized_text`: Sanitized version of the prompt
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let result = shield.scan_prompt("Hello, how are you?").await?;
    ///
    /// if result.is_valid {
    ///     // Safe to send to LLM
    ///     send_to_llm(&result.sanitized_text);
    /// } else {
    ///     // Handle security risk
    ///     log_security_event(&result);
    /// }
    /// ```
    pub async fn scan_prompt(&self, prompt: &str) -> SdkResult<ScanResult> {
        self.scan_with_scanners(prompt, &self.input_scanners).await
    }

    /// Scan an LLM output before showing to the user
    ///
    /// ## Arguments
    ///
    /// * `output` - The LLM response to scan
    ///
    /// ## Returns
    ///
    /// A `ScanResult` containing validation results.
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let llm_response = call_llm(prompt).await?;
    /// let result = shield.scan_output(&llm_response).await?;
    ///
    /// if result.is_valid {
    ///     // Safe to show to user
    ///     display_to_user(&result.sanitized_text);
    /// } else {
    ///     // Handle security risk
    ///     display_error("Response filtered for security reasons");
    /// }
    /// ```
    pub async fn scan_output(&self, output: &str) -> SdkResult<ScanResult> {
        self.scan_with_scanners(output, &self.output_scanners).await
    }

    /// Scan both prompt and output in one call
    ///
    /// ## Arguments
    ///
    /// * `prompt` - The user prompt
    /// * `output` - The LLM response
    ///
    /// ## Returns
    ///
    /// A tuple of (prompt_result, output_result).
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let (prompt_result, output_result) = shield
    ///     .scan_prompt_and_output(prompt, llm_response)
    ///     .await?;
    /// ```
    pub async fn scan_prompt_and_output(
        &self,
        prompt: &str,
        output: &str,
    ) -> SdkResult<(ScanResult, ScanResult)> {
        let prompt_result = self.scan_prompt(prompt).await?;
        let output_result = self.scan_output(output).await?;
        Ok((prompt_result, output_result))
    }

    /// Scan multiple prompts in batch
    ///
    /// Optimized for high-throughput scenarios.
    ///
    /// ## Arguments
    ///
    /// * `prompts` - Slice of prompts to scan
    ///
    /// ## Returns
    ///
    /// Vector of `ScanResult` in the same order as input prompts.
    ///
    /// ## Example
    ///
    /// ```rust,ignore
    /// let prompts = vec!["Hello", "How are you?", "What's the weather?"];
    /// let results = shield.scan_batch(&prompts).await?;
    ///
    /// for (prompt, result) in prompts.iter().zip(results.iter()) {
    ///     println!("{}: {}", prompt, if result.is_valid { "safe" } else { "risky" });
    /// }
    /// ```
    pub async fn scan_batch(&self, prompts: &[&str]) -> SdkResult<Vec<ScanResult>> {
        if prompts.len() > self.config.max_batch_size {
            return Err(SdkError::validation(format!(
                "Batch size {} exceeds maximum {}",
                prompts.len(),
                self.config.max_batch_size
            )));
        }

        // Execute all scans in parallel
        let futures: Vec<_> = prompts
            .iter()
            .map(|prompt| self.scan_prompt(prompt))
            .collect();

        let results: Vec<SdkResult<ScanResult>> = join_all(futures).await;

        // Collect results, returning first error if any
        results.into_iter().collect()
    }

    /// Scan multiple outputs in batch
    ///
    /// ## Arguments
    ///
    /// * `outputs` - Slice of LLM outputs to scan
    ///
    /// ## Returns
    ///
    /// Vector of `ScanResult` in the same order as input outputs.
    pub async fn scan_output_batch(&self, outputs: &[&str]) -> SdkResult<Vec<ScanResult>> {
        if outputs.len() > self.config.max_batch_size {
            return Err(SdkError::validation(format!(
                "Batch size {} exceeds maximum {}",
                outputs.len(),
                self.config.max_batch_size
            )));
        }

        let futures: Vec<_> = outputs
            .iter()
            .map(|output| self.scan_output(output))
            .collect();

        let results: Vec<SdkResult<ScanResult>> = join_all(futures).await;
        results.into_iter().collect()
    }

    // ========================================================================
    // Internal Methods
    // ========================================================================

    /// Execute scanners on input
    async fn scan_with_scanners(
        &self,
        input: &str,
        scanners: &[Arc<dyn Scanner>],
    ) -> SdkResult<ScanResult> {
        if scanners.is_empty() {
            // No scanners configured - pass through
            return Ok(ScanResult::pass(input.to_string()));
        }

        let start = Instant::now();

        // Build pipeline
        let mut pipeline = ScannerPipeline::new();
        for scanner in scanners {
            pipeline = pipeline.add(Arc::clone(scanner));
        }

        // Apply short-circuit if configured
        if let Some(threshold) = self.config.short_circuit_threshold {
            pipeline = pipeline.with_short_circuit(threshold);
        }

        // Execute pipeline
        let result = if self.config.parallel.enabled {
            pipeline
                .execute_parallel(input, &self.vault)
                .await
                .map_err(|e| SdkError::pipeline(e.to_string()))?
        } else {
            pipeline
                .execute(input, &self.vault)
                .await
                .map_err(|e| SdkError::pipeline(e.to_string()))?
        };

        // Combine results
        let combined = ScanResult::combine(result);

        // Add timing metadata
        let elapsed_ms = start.elapsed().as_millis() as u64;
        let final_result = combined.with_metadata("scan_time_ms", elapsed_ms);

        // Check timeout
        if let Some(timeout_ms) = self.config.timeout_ms {
            if elapsed_ms > timeout_ms {
                return Err(SdkError::timeout(timeout_ms));
            }
        }

        Ok(final_result)
    }

    // ========================================================================
    // Configuration Access
    // ========================================================================

    /// Get the current configuration
    pub fn config(&self) -> &ShieldConfig {
        &self.config
    }

    /// Get the number of input scanners
    pub fn input_scanner_count(&self) -> usize {
        self.input_scanners.len()
    }

    /// Get the number of output scanners
    pub fn output_scanner_count(&self) -> usize {
        self.output_scanners.len()
    }

    /// Get input scanner names
    pub fn input_scanner_names(&self) -> Vec<&str> {
        self.input_scanners.iter().map(|s| s.name()).collect()
    }

    /// Get output scanner names
    pub fn output_scanner_names(&self) -> Vec<&str> {
        self.output_scanners.iter().map(|s| s.name()).collect()
    }

    /// Get the vault for advanced usage
    pub fn vault(&self) -> &Vault {
        &self.vault
    }

    /// Clear the vault
    pub fn clear_vault(&self) -> SdkResult<()> {
        self.vault.clear().map_err(|e| SdkError::pipeline(e.to_string()))
    }
}

// Safety: Shield is thread-safe
unsafe impl Send for Shield {}
unsafe impl Sync for Shield {}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_shield_standard() {
        let shield = Shield::standard();
        assert!(shield.is_ok());

        let shield = shield.unwrap();
        assert!(shield.input_scanner_count() > 0);
    }

    #[tokio::test]
    async fn test_shield_strict() {
        let shield = Shield::strict();
        assert!(shield.is_ok());

        let shield = shield.unwrap();
        assert!(shield.input_scanner_count() >= shield.input_scanner_count());
    }

    #[tokio::test]
    async fn test_shield_permissive() {
        let shield = Shield::permissive();
        assert!(shield.is_ok());
    }

    #[tokio::test]
    async fn test_scan_prompt_safe() {
        let shield = Shield::permissive().unwrap();
        let result = shield.scan_prompt("Hello, how are you?").await;

        assert!(result.is_ok());
        let result = result.unwrap();
        assert!(result.is_valid);
        assert!(result.risk_score < 0.5);
    }

    #[tokio::test]
    async fn test_scan_prompt_with_secret() {
        let shield = Shield::standard().unwrap();
        let result = shield
            .scan_prompt("My AWS key is AKIAIOSFODNN7EXAMPLE")
            .await;

        assert!(result.is_ok());
        let result = result.unwrap();
        assert!(!result.is_valid);
        assert!(result.risk_score > 0.5);
    }

    #[tokio::test]
    async fn test_scan_output() {
        let shield = Shield::standard().unwrap();
        let result = shield
            .scan_output("Here is the information you requested.")
            .await;

        assert!(result.is_ok());
    }

    #[tokio::test]
    async fn test_scan_batch() {
        let shield = Shield::permissive().unwrap();
        let prompts = vec!["Hello", "How are you?", "What's the weather?"];
        let results = shield.scan_batch(&prompts).await;

        assert!(results.is_ok());
        let results = results.unwrap();
        assert_eq!(results.len(), 3);
    }

    #[tokio::test]
    async fn test_scan_batch_exceeds_limit() {
        let shield = Shield::builder()
            .with_preset(Preset::Permissive)
            .with_max_batch_size(2)
            .build()
            .unwrap();

        let prompts = vec!["1", "2", "3"];
        let result = shield.scan_batch(&prompts).await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_scanner_names() {
        let shield = Shield::standard().unwrap();

        let input_names = shield.input_scanner_names();
        assert!(!input_names.is_empty());

        let output_names = shield.output_scanner_names();
        assert!(!output_names.is_empty());
    }

    #[tokio::test]
    async fn test_vault_operations() {
        let shield = Shield::permissive().unwrap();

        // Vault should be accessible
        let vault = shield.vault();
        assert!(vault.is_empty());

        // Clear should work
        let result = shield.clear_vault();
        assert!(result.is_ok());
    }
}
